### ISSUE TITLE
TypeError: HidefixBackendEntrypoint.open_dataset() got an unexpected keyword argument 'group'

### ISSUE THREAD

**Issue by weiji14** [gauteh/hidefix#24]
Hi there,

Been following your work so far and it is looking really promising! Just reporting this 'bug' when attempting to use the 'group' parameter in [`xr.open_dataset`](https://docs.xarray.dev/en/v2023.06.0/generated/xarray.open_dataset.html).

### Using hidefix

Download this sample HDF5 file: https://github.com/suzanne64/ATL11/raw/125ee1a653d78e6b86864b35c9d0fcfd72d64a85/ATL11_test_case/ATL11_078805_0304_02_v002.h5, and read group `pt2/corrected_h`.

```python
import hidefix
import xarray as xr

ds: xr.Dataset = xr.open_dataset(
    filename_or_obj="ATL11_078805_0304_02_v002.h5",
    engine="hidefix",
    group="pt2/corrected_h",
)
print(ds)
```
produces this error:
```python-traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File <timed exec>:2

File /srv/conda/envs/notebook/lib/python3.10/site-packages/xarray/backends/api.py:1012, in open_mfdataset(paths, chunks, concat_dim, compat, preprocess, engine, data_vars, coords, combine, parallel, join, attrs_file, combine_attrs, **kwargs)
   1009     open_ = open_dataset
   1010     getattr_ = getattr
-> 1012 datasets = [open_(p, **open_kwargs) for p in paths]
   1013 closers = [getattr_(ds, "_close") for ds in datasets]
   1014 if preprocess is not None:

File /srv/conda/envs/notebook/lib/python3.10/site-packages/xarray/backends/api.py:1012, in <listcomp>(.0)
   1009     open_ = open_dataset
   1010     getattr_ = getattr
-> 1012 datasets = [open_(p, **open_kwargs) for p in paths]
   1013 closers = [getattr_(ds, "_close") for ds in datasets]
   1014 if preprocess is not None:

File /srv/conda/envs/notebook/lib/python3.10/site-packages/xarray/backends/api.py:570, in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, inline_array, chunked_array_type, from_array_kwargs, backend_kwargs, **kwargs)
    558 decoders = _resolve_decoders_kwargs(
    559     decode_cf,
    560     open_backend_dataset_parameters=backend.open_dataset_parameters,
   (...)
    566     decode_coords=decode_coords,
    567 )
    569 overwrite_encoded_chunks = kwargs.pop("overwrite_encoded_chunks", None)
--> 570 backend_ds = backend.open_dataset(
    571     filename_or_obj,
    572     drop_variables=drop_variables,
    573     **decoders,
    574     **kwargs,
    575 )
    576 ds = _dataset_from_backend_dataset(
    577     backend_ds,
    578     filename_or_obj,
   (...)
    588     **kwargs,
    589 )
    590 return ds

TypeError: HidefixBackendEntrypoint.open_dataset() got an unexpected keyword argument 'group'
```

### Using h5netcdf engine (expected results)

```python
ds: xr.Dataset = xr.open_dataset(
    filename_or_obj="ATL11_078805_0304_02_v002.h5",
    engine="h5netcdf",
    group="pt2/corrected_h",
)
print(ds)
```
produces
```
<xarray.Dataset>
Dimensions:                  (cycle_number: 2, ref_pt: 1404)
Coordinates:
  * cycle_number             (cycle_number) int64 3 4
  * ref_pt                   (ref_pt) float64 6.016e+05 6.016e+05 ... 6.086e+05
Data variables:
    delta_time               (ref_pt, cycle_number) timedelta64[ns] ...
    h_corr                   (ref_pt, cycle_number) float64 ...
    h_corr_sigma             (ref_pt, cycle_number) float64 ...
    h_corr_sigma_systematic  (ref_pt, cycle_number) float64 ...
    latitude                 (ref_pt) float64 ...
    longitude                (ref_pt) float64 ...
    quality_summary          (ref_pt, cycle_number) float64 ...
```

## Library versions

Using hidefix=0.6.5. Other library versions from `xr.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 5.10.167-147.601.amzn2.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.0
libnetcdf: 4.9.2

xarray: 2023.7.0
pandas: 1.5.1
numpy: 1.23.5
scipy: 1.9.3
netCDF4: 1.6.4
pydap: None
h5netcdf: 1.1.0
h5py: 3.9.0
Nio: None
zarr: 2.15.0
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: None
dask: 2022.11.0
distributed: 2022.11.0
matplotlib: 3.6.2
cartopy: 0.21.1
seaborn: 0.12.1
numbagg: None
fsspec: 2023.6.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 67.7.2
pip: 22.3.1
conda: None
pytest: 7.2.0
mypy: None
IPython: 8.6.0
sphinx: 4.5.0
```

Would be keen to help implement the 'group' reading feature somehow, but I might need a lot of pointers on how to handle the Rust bindings to figure things out!

**Comment by gauteh** [comment]
Great! I think that should be possible without too much effort, though I have not tested it. The first step would be to create a rust test where the group is opened and indexed+read, e.g. like: https://github.com/gauteh/hidefix/blob/main/tests/read_double.rs#L8

This is probably best done by opening the dataset using hdf5-rs, finding the group, getting the Dataset object and indexing in. If that is possible the next step would be to add a method here: https://github.com/gauteh/hidefix/blob/main/src/python/mod.rs#L37 for opening a dataset in a group (or following a path or something).

Does that sound possible? I would be happy to help further.

**Comment by weiji14** [comment]
Cool, thanks for the pointers! Let me give this a shot, this is actually my first time diving into Rust, but I'm been meaning to learn for a while, and this will be a nice task to work on. I've got a dev environment for hidefix setup already, and just need some time to walk through the codebase. Will give you a shout if I need any help!

**Comment by gauteh** [comment]
Sounds good, good luck. I think this is a fun project to learn rust on, you can play with optimizations, and most of them can be done in idiomatic rust without any loss over optimized C (e.g. duffs-device in shuffle.rs).

**Comment by weiji14** [comment]
Quick question, would it be possible to set the 'group' to read by modifying the code somewhere here:

https://github.com/gauteh/hidefix/blob/0e637149e159164ea23c2e244080399bd34f8cb5/src/idx/index.rs#L51-L56

I.e. change the hardcoded root group `"/"` at L56 to some String reference? My understanding is that I could do something like `group.unwrap_or(String::from("/"))`, but still figuring out how to set an optional 'group' parameter/argument into the `index` and `index_file` function first. Let me read the Rust book a bit more to understand the language syntax first.

**Comment by gauteh** [comment]
Yes, that should work. I think the best way to do it would be as outlined in https://github.com/gauteh/hidefix/issues/25, create an additional group layer between index and datasets. Then if you try to access datasets on the index struct it should deref to the root group. But you can definitely start by adding a group argument, or creating a new method that also takes a group.

**Comment by gauteh** [comment]
This can now be fixed in the python bindings.

**Comment by gauteh** [comment]
Should be fixed in #36
